Source/JavaScriptCore/wasm/js/WebAssemblyBuiltin.cpp

```
@@ -0,0 +1,825 @@
+        && elementType.type.as<Wasm::PackedType>() == Wasm::PackedType::I16;
+}
+
+#define PASTE(x, y) x##y
+#define CONCAT(x, y) PASTE(x, y)
+
+// The naming scheme. Changes should be matched in InPlaceInterpreter.asm
+// macros wasmBuiltinCallTrampoline and defineWasmBuiltinCallTrampoline.
+
+#define BUILTIN_TRAMPOLINE(setName, builtinName) CONCAT(wasm_builtin_trampoline__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_WASM_ENTRY(setName, builtinName) CONCAT(wasm_builtin__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_JS_ENTRY(setName, builtinName) CONCAT(wasm_builtinJS__, BUILTIN_FULL_NAME(setName, builtinName))
+#define BUILTIN_IMPLEMENTATION(setName, builtinName) CONCAT(wasm_builtin_impl__, BUILTIN_FULL_NAME(setName, builtinName))
+
+#define DEFINE_BUILTIN_IMPLEMENTATION(setName, builtinName, ...) \
+    static ALWAYS_INLINE EncodedJSValue BUILTIN_IMPLEMENTATION(setName, builtinName)(__VA_ARGS__)
+
+#define DEFINE_BUILTIN_IMPLEMENTATION_I32(setName, builtinName, ...) \
+    static ALWAYS_INLINE int32_t BUILTIN_IMPLEMENTATION(setName, builtinName)(__VA_ARGS__)
+
+#define DECLARE_BUILTIN_TRAMPOLINE(setName, builtinName) \
+    extern "C" EncodedJSValue BUILTIN_TRAMPOLINE(setName, builtinName)()
+
+// A host wasm function that returns an externref (EncodedJSValue)
+#define DEFINE_WASM_BUILTIN_HOST_FUNCTION(setName, builtinName, ...) \
+    extern "C" EncodedJSValue SYSV_ABI BUILTIN_WASM_ENTRY(setName, builtinName)(__VA_ARGS__) REFERENCED_FROM_ASM WTF_INTERNAL; \
+    EncodedJSValue SYSV_ABI BUILTIN_WASM_ENTRY(setName, builtinName)(__VA_ARGS__)
+
+#define THROW_EXCEPTION \
+    do { \
+        JSObject* error = createJSWebAssemblyRuntimeError(globalObject, globalObject->vm(), Wasm::ExceptionType::IllegalArgument); \
+        throwException(globalObject, scope, error); \
+        return encodedJSValue(); \
+    } while (false)
+
+#define IF_EXCEPTION_RETHROW_AND_RETURN(scope) \
+    if (scope.exception()) \
+        THROW_EXCEPTION;
```

This doesn't seem right for the JS API. We should just rethrow the error with `RETURN_IF_EXCEPTION(scope, { });`

---

I renamed the macro to a more descriptive `IF_EXCEPTION_RETHROW_AS_ILLEGAL_ARGUMENT`. This is the typical exception regrowing pattern, when a piece of logic catches an exception and rethrows it as a different one to expose a consistent exception surface to its callers. In our case we normally throw an exception saying "Illegal argument" when an argument is bad, and it should be the same no matter how we discover that the argument is bad.